### ISSUE TITLE
IANA protocol register

### ISSUE THREAD

**Issue by 6543** [HelloZeroNet/ZeroNet#2100]
we should register the zero:// handler

https://www.iana.org/form/protocol-assignment

**Comment by filips123** [comment]
See also #2087.

It would be good to standardize and registered almost all parts of ZeroNet. This includes standardization of ZeroNet protocol (as new RFC) and `zero://` protocol, `.zeronet` TLD and port `43110` (at IANA).

Problem is that some of that standardizations require to register a company and pay a large amount of money. For example, to register official TLD you need to pay $185000.

**Comment by 6543** [comment]
so do it piece by piece?
if TLD is to expencive ... start with RFC and port 43110?

PS: spell check loves me :D so RFC isn't the best task for me

**Comment by Merith-TK** [comment]
I would be willing to donate power to mine menero to do this

**Comment by 6543** [comment]
@Merith-TK this is an good idear i created #2101

**Comment by filips123** [comment]
It might be possible to register `.zeronet` domain as "Special-Use Domain". I don't know if there is also a requirement to pay for that but it is based on RFCs.

For example, TLD `.onion` is registered for special use with [RFC 7686](https://tools.ietf.org/rfc/rfc7686.txt).

Same should be done (if possible) for `.zeronet`. It should not be hard to write that RFC. For example `.onion`'s RFC contains only 7 pages and actual RFC text is only on 4 pages.

**Comment by filips123** [comment]
Here are also my thoughts about different ways about accessing ZeroNet and implementing them (see #2231 and #2087). I commented this here because there are currently quite some ZeroNet issues about DNS/protocols/schemes... and I think that this one is currently the most suitable for this type of comment.

---

## Ways of accessing ZeroNet

### zero:// scheme

* Simpler to implement than `z*://` schemes.
* A bit harder to implement than `.zeronet` TLD.
* Is almost schematically valid.
* But it in conflict with support for protocols other than HTTP and ZeroNet communication protocol.
* Does not allow using public ZeroNet instances.

### z*:// schemes

* Could be `zhttp://` for HTTP, `zhttps://` for HTTPS, `zftp://` for FTP...
* Harder to implement as each custom scheme would need to be handled separatly in the browser and every program you want to use it (local programs, CURL, API clients...).
* More schematically valid as it correctly separates different access protocols.
* Still does not allow using public ZeroNet instances.

### .zeronet TLD

* Simpler to implement (see the second part of this comment).
* Can be used in any program without too mich changes (also see second part of comment).
* Schematically valid as it separates all ZeroNet related sites with custom TLD.
* Allows using public ZeroNet instances with `https://address.public.zeronet.com`.
* Similar way is also used in Tor (`.onion`).

## Ways of implementing this

### Extensions

* Easy for users to install and use.
* Not hard to do as extension can actually install/use PAC file.
* But you have to pay for Google and Microsoft store (or hire some existing developer to publish it for you).
* Not well supported on mobile (but you can use Firefox for Android which support extensions).

### PAC file

* A bit harder for users.
* Very easy to implement.
* Not supported in all browsers but it can also be handled by extensions.
* Can also be a global setting (would be useful for ZeroNet applications outside the web).

### Proxy / Local DNS / Local DoH

* Forward all traffic through DNS server which then decides if it is handled by ZeroNet or clearnet.
* Can also be a global setting.
* Can be easy to implement if you already have DNS server library for Python.

### Local system program/app / System-wide configuration

* Can be used for system-wide settings.
* Can be used on macOS, mobile or systems without other easy ways systems.
* Can be actually the only wrapper around other methods (PAC file, DNS server).

---

For the way of accessing ZeroNet, I would most like `.zeronet` TLD as it is not hard to implement and also serves its purpose. This would also allow accessing public ZeroNet instances `https://address.public.zeronet.com` and even directly accessing ZeroNet sites with DNSLink and CNAME/A record (part of this is #2214 but the other half still needs to be implemented - can be probably implemented by @imachug in "Road to better sandbox" series).

For the way of implementing this, I'd start with PAC file as it is also the easiest. Then it should be expanded to browser extensions (which can actually just be wrappers around PAC files) and then custom DNS server and system programs. But first of all, the ZeroNet itself should support handling all this things. This is something that @imachug is currently working on.

**Comment by purplesyringa** [comment]
A small con to proxy-based solution: the proxy has to be always online which is not always possible. Same with DNS, though modern OSes usually have fallbacks. Still, configuring global DNSes might get troublesome if you are using dnsmasq or such already or if you want ZeroNet installation to be easy.

**Comment by purplesyringa** [comment]
> A bit harder for users.

There's a small problem with this. Unfortunately, it looks like it's *a lot* harded for users.

Before: you download a ZIP, extract it and run ZeroNet.exe. Hooray, you can now browse ZeroNet.

After: you download a ZIP, extract it and run ZeroNet.exe. ZeroNet asks you to install some random extension. You click the link and press "Install" button. You wait for the extension to get installed, open its settings, notice "PAC file", press "Choose" button. You then notice that you have no idea what PAC you're looking for. You press Esc, go back to ZeroNet homepage, look at the path and decide to take it down. You return to the settings page, type in the path and press OK. Hooray, ZeroNet now works.

**Comment by filips123** [comment]
@imachug It is not so hard to install an extension. The extension can already pack proxy inside it so normal users don't need to configure it manually. See [`proxy.onRequest`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/proxy/onRequest) for Firefox API but I think that Chromium also supports similar thing.

So it would actually look something like this: you download a ZIP, extract it and run ZeroNet.exe. ZeroNet asks you to install some random extension. You click the link and press "Install" button. You wait for the extension to get installed. Hooray, ZeroNet now works.

Just PAC file is another thing, but it doesn't need an extension. It should be supported first, but soon after that, the extension should be made.

**Comment by purplesyringa** [comment]
Welp, that means that we'll have to make our own extension. Whilst Firefox allows side-loading extensions, Chrome and Edge don't, Safari doesn't support WebExtensions, and mobile Chrome doesn't support extensions at all. Oopsie.

**Comment by filips123** [comment]
That's why I said that extension should be **one of** the supported solutions. Because most users (desktop) can use it and it is not very hard to do, it should be one of the first implementations.

For mobile users, ZeroNet is actually already not so optimized, so this won't be such drawback. But in the future, there could be some native Android app to do this (same for macOS). Currently, you can use Firefox or another Android browser with extensions.

And even if companies (Microsoft and Google) require you to pay for to register developer account, you may still be able to publish it for free. I'm sure that there is at least one active and trusted ZeroNet user/developer which has a developer account and would be happy to publish extensions for you.

**Comment by Merith-TK** [comment]
Going to but in here, CHROME DOES, it just screams at startup saying
"developer extensions are enabled" and any button disables them

on chrome, go to `chrome://extensions` and click the "developer mode" and
then load Unpacked extension, just point that to the root directory of the
extension
![image](https://user-images.githubusercontent.com/10422110/67246225-bbde2f80-f412-11e9-9039-f591fcbba24c.png)

**Comment by Merith-TK** [comment]
![image](https://user-images.githubusercontent.com/10422110/67246259-cb5d7880-f412-11e9-9881-affe7267c9bc.png)

**Comment by purplesyringa** [comment]
Right, but that's really difficult for an average user if we want a big user base.

**Comment by Merith-TK** [comment]
How is it difficult? Hell, setup VPN installer has a method of sideloading
the extention so all it would take is making a EXE to sideload it

**Comment by purplesyringa** [comment]
Sideloading an extension requires a Google developer account and uploading the extension to store. It also means that Google can revoke the extension whenever they want.

**Comment by Merith-TK** [comment]
That is wrong actualy, you just need chrome, 
Uploading to the store, yes, you need a dev account. but for making a unpacked extension, you just need chrome to sign it, 
look at this as a example you have git clone it, and load unpacked extension

https://github.com/goldenratio/zeronet-protocol-crx

**Comment by purplesyringa** [comment]
> But for local ZeroNet instances, only .zeronet should be used.

Anyone with a google developer account here? Anyways, Google centralization doesn't look like a good idea to me. Look at what they did to AdNauseam.

**Comment by Merith-TK** [comment]
Okay, WE DONT NEED the google store, how i would have this, is as follows

have a zite that contains the instructions to enable the extension, and have the zite come with the unpacked extension we need, so that way updating is easy as pie. 

1) make extension
2) upload unpacked extenstion to zite
3) have zite contain instructions to enable.

I could probably make the zite instructions, (in basic HTML/CSS), but i dont know enough to actually make a real distrobuted Zite yet

**Comment by purplesyringa** [comment]
That would yield a "This extension is installed in developer mode" notification every time.

**Comment by Merith-TK** [comment]
Which I have already stated would be one of the drawbacks to it.

**Comment by purplesyringa** [comment]
Right. That's why I think it should only be thought of as the worst solution. We should try to avoid that as much as we can.

**Comment by Merith-TK** [comment]
Yea, maybe something like Ungoogled Chromium or Brave support, or some third party browser with a plugin, those should allow third party plugins... i think

**Comment by filips123** [comment]
@Merith-TK I will maybe start working on custom Firefox/Tor Browser bundle with support for ZeroNet. But  I don't promise anything because I don't have much time.

But really, we should first check if there is any existing extension developer which can publish an extension to Web Store so there won't be any need for that "hacky solutions". But they would probably be needed (and implemented) some time later.

**Comment by Merith-TK** [comment]
for now, here is what i would reccomend using this pac file, as it works. Maybe an extension that adds this pac file?
```pac
function FindProxyForURL(url, host) {

 if (shExpMatch(host, "*.bit"))
  return "PROXY 127.0.0.1:43110";

 if (shExpMatch(host, "*.zeroid"))
  return "PROXY 127.0.0.1:43110";

 if (shExpMatch(url, "http://zero/*"))
  return "PROXY 127.0.0.1:43110";

 if (shExpMatch(url, "zero://*"))
  return "PROXY 127.0.0.1:43110";

 if (shExpMatch(url, "http://www.zero.com/*"))
  return "PROXY 127.0.0.1:43110";

 return "DIRECT";
}
```

![image](https://user-images.githubusercontent.com/10422110/67320907-b42a9380-f4c3-11e9-9672-75acaf9688ed.png)

**Comment by filips123** [comment]
@Merith-TK It doesn't support `.zeronet` TLD. Also, because support for all these unofficial ways would not be really needed, this should be enough:

```js
function FindProxyForURL(url, host) {
  if (shExpMatch(host, '*.zeronet')) {
    return 'PROXY localhost:43110'
  }

  return 'DIRECT'
}
```

**Comment by Merith-TK** [comment]
except that does not factor in `.bit` domains, or other domain extensions

**Comment by filips123** [comment]
@Merith-TK This is the point. `.bit` domains should not be handled separately but under special `.zeronet` TLD. This will make them much easier to handle than each one separately and would also prevent collisions with normal `.bit` domains.

* Addresses: `http://1HeLLo4uzjaLetFx6NH3PMwFP3qbRbTf3D.zeronet`
* Namecoin domains: `http://Talk.ZeroNetwork.bit.zeronet`
* ENS domains: `http://example.bit.zeronet`
* etc.

See https://github.com/HelloZeroNet/ZeroNet/issues/2087#issuecomment-509829391 for derails what needs to be done to implement this.

**Comment by Merith-TK** [comment]
While that looks like it might work, 
it throws a err with zeronet
![image](https://user-images.githubusercontent.com/10422110/67323345-2badf200-f4c7-11e9-87a6-cd30cd13d54e.png)

**Comment by filips123** [comment]
@Merith-TK See [#2087 (comment)](https://github.com/HelloZeroNet/ZeroNet/issues/2087#issuecomment-509829391) for details what needs to be done to implement this:

> Then the ZeroNet code needs to be updated. There are currently a few problems which need to be resolved:
> 
> * Server will throw `Error 403: Invalid host`. The server also needs to accept hosts that end with `.zeronet` (maybe there should also be some additional security checks). There is already [4ffd642](https://github.com/HelloZeroNet/ZeroNet/commit/4ffd6427323945fcba734f60209c0d0dbdb7b4d5) but it needs to be enabled by default and with more features (support for `.zeronet`).
> * Code needs to be updated to ignore `.zeronet` at the end of the address. If not, the site will not be found.
> * Addresses will turn into lowercase, so they will become invalid. There needs to be some solution for that. There is also PR #2053 for multiple cryptographies and some of them might support case insensitive addresses.
> * There could be problems with UiPassword or Multiuser plugin. Users who use those plugins will need to input password for every site they visit (because cookies will be only for one domain). See my issue #2050 for better user management.

**Comment by Merith-TK** [comment]
Alright, those do make sense, but the pac file stuff i sent actually works. So it could be a implementation until we can get it fully operational

**Comment by filips123** [comment]
@Merith-TK I agree with you. But I'm talking about "official solution", not about "one of the community versions". Yes, *currently* your PAC file is correct. But when ZeroNet implements official proxy/domain support, my version should be used.

**Comment by Merith-TK** [comment]
Alright, i agree with you on that,

**Comment by purplesyringa** [comment]
A brief overview on what we could do on different browsers:

- Chrome desktop: extension (blocked cause we need a developer), PAC file (global, Windows-only)
- Firefox desktop: extension, PAC file, local proxy
- Edge desktop: extension (blocked cause we need a developer)
- Safari desktop: extension
- Chrome mobile: none?
- Firefox mobile: extension
- Safari mobile: none?

So, extensions are supported by most browsers but Chrome and Edge will make them a pain if we don't use a developer acccount. As for mobile browsers, we could make a dedicated app for them.

**Comment by d14na** [comment]
@imachug when i search the chrome store for __"Zeronet"__, all that comes up is [Free the Wiki](https://chrome.google.com/webstore/detail/free-the-wiki/jkhhfiobfgkjkhaamiipbegdpeaoffpm). i'm a bit surprised that no one has published a browser extension for zeronet. may i ask why not?

does the __SOURCE CODE__ for a Zeronet extension (or similar -- thinking TOR) already exist? if all you need is a developer account (iirc cost about $25 bucks) then I wouldn't have a problem publishing the extension under our __d14nauth[at]gmail__ account. don't have a __d14nauth[at]msn__ dev account, but that really shouldn't be a problem

**Comment by Merith-TK** [comment]
@d14na 
https://github.com/goldenratio/zeronet-protocol-crx
I actually think google themselves took it down...

**Comment by purplesyringa** [comment]
> does the SOURCE CODE for a Zeronet extension (or similar -- thinking TOR) already exist?

There's an outdated version but we'll have to modify that for the new sandbox.

> I wouldn't have a problem publishing the extension under our d14nauth[at]gmail account

It'd be cool. I'd use that account for some compatibility tests, I guess I'll need it soon.

> I actually think google themselves took it down...

Let's ping @goldenratio.

**Comment by filips123** [comment]
> So, extensions are supported by most browsers but Chrome and Edge will make them a pain if we don't use a developer acccount.

There are currently two ZeroNet browser extensions.

One is by @goldenratio ([here](https://github.com/goldenratio/zeronet-protocol-crx)) and it uses PAC file so URLs are correctly rewritten. It was published to Chrome Web Store once ago, but it appears that it was taken down. It is also very outdated and does not support other browsers.

The newer extension is by @sprite-1 ([here](https://github.com/sprite-1/zeronetloader)) and it doesn't use PAC file so tabs are just redirected. It not discontinued and supports multiple browsers. However, However, it's author also don't have Google developer account so it is not published to Chrome Web Store.

In my opinion, these two extensions should be merged with each other to provide the best things of both. Then that extension should be published as an official ZeroNet browser extension.

> As for mobile browsers, we could make a dedicated app for them.

For mobile, there could be a dedicated app for ZeroNet.

One possible way how it could function is that it could somehow add ZeroNet as a proxy to system-level (so it would be possible to use it in chrome).

Another possible way is to create a browser app which would only add ZeroNet proxy to it. I think that it is not very hard to implement a browser with [GeckoView](https://github.com/mozilla/geckoview) and [Android Components](https://github.com/mozilla-mobile/android-components) or to [Firefox Preview](https://github.com/mozilla-mobile/fenix) fork. The reason why using GeckoView and not default system WebView is that we shouldn't be another "decentralized" project which uses browser engine from Google.

---

Also, I think that, at least for some time, ZeroNet should support both normal access from `https://127.0.0.1:43110` and proxied access with extensions. The reason is that some users don't want (or can't) set up "proxy" (or whatever will be used) and it is easier to just use current ZeroNet accessing way.

**Comment by purplesyringa** [comment]
> In my opinion, these two extensions should be merged with each other to provide the best things of both.

I think we should rewrite the extensions. They're really simple but we'll need to modify them when we change the way sandbox works. I also have some proxy-related ideas.

**Comment by purplesyringa** [comment]
A small update: PAC file won't work because websocket connections aren't affected. Extensions are required.

**Comment by filips123** [comment]
@imachug Yes but... don't extensions actually use the same browser APIs as PAC file?

**Comment by purplesyringa** [comment]
Nope, the APIs are different. PAC files just define a few functions, whilst extensions have access to `proxy` API.

**Comment by filips123** [comment]
Can `proxy` APIs affect WebSockets?

**Comment by purplesyringa** [comment]
My test extension shows that it can do that.

**Comment by filips123** [comment]
Nice! But why PAC files don't work?

**Comment by purplesyringa** [comment]
¯\\\_(ツ)\_/¯ By definition. PAC files were intended to only proxy a few requests, then websockets were invented but no one wanted to modify PAC behavior. Anyways, I don't think that using webextensions instead of PAC files is a problem.

**Comment by d14na** [comment]
> https://github.com/goldenratio/zeronet-protocol-crx
> I actually think google themselves took it down...

me and google don't get a long for this very reason. long story short, they're the ones who drove me to Bitcoin in 2012 (still kinda bitter) and I haven't looked back.

> In my opinion, these two extensions should be merged with each other to provide the best things of both. Then that extension should be published as an official ZeroNet browser extension.

sounds good to me!

> For mobile, there could be a dedicated app for ZeroNet.

still <s>working on</s> struggling with [Zeronet Explorer](http://127.0.0.1:43110/zeronetexplorer.bit)

---

i probably won't be able to contribute much to the code, (1) i've never done an extension, (2) i have a FULL plate already; but if you send me the .zip, publishing doesn't appear to be an issue (there's a short review process), my dev account is in good shape and any docs needed are in order

**Comment by purplesyringa** [comment]
@d14na Can you please take a look at the extension I published [here](https://github.com/HelloZeroNet/ZeroNet/pull/2231#issuecomment-546830889)?

**Comment by ghost** [comment]
I got here from #83 looking for a way to do http-less zeronet. I see there are a couple of plugins but it's not clear which one will give me `zero://` scheme redirects out of the box. Regardless of whether or not it's _the correct way_ I feel using `zero://` is the best option because it's easy to remember, easy to type and leaves room for the underlying protocol to change over time. IIRC this part of the URI is actually called a _scheme_ not a protocol in case that helps remove some of the "stink" I saw referenced elsewhere. The rest is an implementation detail and should be greenfield not standard.

On growing the userbase... sell people the idea of `.bit` domains for sale like it's the beginning of the WWW all over. Appending anything to `.bit` ruins the sexiness of obtaining 3 and 4 letter domain names.
